pkg/edit/builtins.go

```diff
@@ -563,60 +649,137 @@ func categorizeAlnum(r rune) int {
+func skipWordRight(categorize categorizer, buffer string, pos int) int {
+	if pos == len(buffer) {
+		return pos
+	}
+
+	r, _ := utf8.DecodeRuneInString(buffer[pos:])
 	cat := categorize(r)
+	return skipCatRight(categorize, cat, buffer, pos)
+}
+
+// Skip whitespace right from pos in buffer and return that position.
+func skipWsRight(categorize categorizer, buffer string, pos int) int {
+	return skipCatRight(categorize, 0, buffer, pos)
+}
+
+func skipCatRight(categorize categorizer, cat int, buffer string, pos int) int {
+	right := strings.TrimLeftFunc(buffer[pos:], func(r rune) bool {
+		return categorize(r) == cat
+	})
+
+	return len(buffer) - len(right)
+}
+
+// Move the dot left one word, using the word flavor described by the
+// categorizer.
+func moveDotLeftGeneralWord(categorize categorizer, buffer string, dot int) int {
+	// skip trailing whitespaces left of dot
+	pos := skipWsLeft(categorize, buffer, dot)
 
 	// skip this word
-	skipCat(cat)
+	pos = skipWordLeft(categorize, buffer, pos)
 
-	return len(left)
+	return pos
 }
 
 // Move the dot right one word, using the word flavor described by the
 // categorizer.
 func moveDotRightGeneralWord(categorize categorizer, buffer string, dot int) int {
-	right := buffer[dot:]
-	skipCat := func(cat int) {
-		right = strings.TrimLeftFunc(right, func(r rune) bool {
-			return categorize(r) == cat
-		})
-	}
-
 	// skip leading whitespaces right of dot
-	skipCat(0)
+	pos := skipWsRight(categorize, buffer, dot)
 
 	// check whether any whitespace was skipped; if whitespace was
 	// skipped, then dot is already successfully moved to next
 	// non-whitespace run
-	if dot < len(buffer)-len(right) {
-		return len(buffer) - len(right)
+	if dot < pos {
+		return pos
 	}
 
 	// no whitespace was skipped, so we still have to skip to the next word
 
-	// get category of first rune
-	r, _ := utf8.DecodeRuneInString(right)
-	cat := categorize(r)
 	// skip this word
-	skipCat(cat)
+	pos = skipWordRight(categorize, buffer, pos)
 	// skip remaining whitespace
-	skipCat(0)
+	pos = skipWsRight(categorize, buffer, pos)
 
-	return len(buffer) - len(right)
+	return pos
+}
+
+// Transpose the words around the cursor, using the word flavor described
+// by the categorizer.
+func transposeGeneralWord(categorize categorizer, buffer string, dot int) (string, int) {
+	if strings.TrimFunc(buffer, func(r rune) bool { return categorize(r) == 0 }) == "" {
+		// buffer contains only whitespace
+		return buffer, dot
+	}
+
+	// after skipping whitespace, find the end of the right word
+	pos := skipWsRight(categorize, buffer, dot)
+	var rightEnd int
+	if pos == len(buffer) {
+		// there is only whitespace to the right of the dot
+		rightEnd = skipWsLeft(categorize, buffer, pos)
+	} else {
+		rightEnd = skipWordRight(categorize, buffer, pos)
+	}
+	// if the dot started in the middle of a word, 'pos' is the same as dot,
+	// so we should skip word characters to the left to find the start of the
+	// word
```

Comment:
I see you added this as a response to my question, thanks. The comment feels a bit out of place because it's responding to a particular doubt about why this is not equal to `pos`, which the prospective reader does not necessarily have. I think it's probably better just to remove this comment, the code is actually clear enough.